clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java

```diff
@@ -582,13 +582,8 @@ private void postProcessAndValidateIdempotenceConfigs(final Map<String, Object>
 
             final int inFlightConnection = this.getInt(MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION);
             if (MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION_FOR_IDEMPOTENCE < inFlightConnection) {
-                if (userConfiguredIdempotence) {
-                    throw new ConfigException("Must set " + MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION + " to at most 5" +
-                        " to use the idempotent producer.");
-                }
-                log.warn("Idempotence will be disabled because {} is set to {}, which is greater than 5. " +
-                    "Please note that in v4.0.0 and onward, this will become an error.", MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION, inFlightConnection);
-                shouldDisableIdempotence = true;
+                throw new ConfigException("To use the idempotent producer, " + MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION +
```

Comment:
How about updating the validator from `atLeast(1)` to `between(1, 5)`, so we don't need the check here?

```java
                                .define(MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION,
                                        Type.INT,
                                        5,
                                        between(1, 5),
                                        Importance.LOW,
                                        MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION_DOC)
```

Reply:
Hmmm… This logic indeed covers the case where `enable.idempotence` is set to `true`, but we don’t want this constraint if `enable.idempotence` is set to `false`, right?

Reply:
IIUC, broker side only contains at most 5 batches. Event if the `enable.idempotence` is `false`, the maximum `MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION` can't be more than `5`.

https://github.com/apache/kafka/blob/af53758746485daa06c4d4366036c04f7f9f0e63/storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateEntry.java#L29-L36

https://github.com/apache/kafka/blob/af53758746485daa06c4d4366036c04f7f9f0e63/storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateEntry.java#L107-L110